### ISSUE TITLE
Offline read access to resource definition.

### ISSUE THREAD

**Issue by eirrgang** [radical-cybertools/radical.pilot#2668]
The complete data represented by the https://radicalpilot.readthedocs.io/en/stable/apidoc.html#radical.pilot.PilotDescription.resource is not easily programmatically accessible to clients.

The resource definition can come from multiple places, has a schema that is either loosely specified or poorly documented (see also #2588), and is not easily accessed (at least without an active Session).

### Check resource accessibility

When the *resource* is being managed abstractly (symbolically) by a script, it would be helpful to be able to discover the connection method to the resource. This would allow scripted debugging of connection issues, as well as extra checks for environment suitability (e.g. checking for installed software or software versions, or even preconfiguring Python virtual environments before submitted large HPC job scripts).

### Discover filesystem endpoint

Staging large data _can_ be done through the `Pilot.stage_in`, but may result in wasted compute allocation if a large pilot job starts while data is still transferring.

Large data should be staged before the Pilot is submitted, and preferably without requiring the resources of an active Session. If the `radical.utils.Url` of the filesystem endpoint could be retrieved, more basic RCT/Saga utilities could be used to stage data before allocating HPC resources.

Even if there were a canonical / blessed way to retrieve the string representation of the filesystem endpoint while offline, client scripts could find ways to programmatically and consistently access the appropriate remote filesystem and path to stage data before or after a Session (or even from separate client-side utility scripts that manage RP inputs and artifacts). See also #2590.

## Possible fixes
* Support `radical.pilot.session.Session.get_resource_config()` as a `@classmethod`, usable without an active Session.
* Provide a simple _public_ utility to aggregate a resource definition from its various sources.
* Document the stable parts of the resource schema and provide getters.
* Provide a query tool to get arbitrary fields or fail with an appropriate exception.

See also #2588 

## Deferred

It would be great to have API for setting or updating resources known to the Session, but I intend to explicitly exclude that topic from this feature request to limit the scope of the issue.

**Comment by eirrgang** [comment]
@mturilli asked me to document this feature request. We were discussing how best to separate the abstraction of execution management from data management. I observed that the "interface" between the two spaces would be clearer and cleaner if it were easier to consistently identify the same filesystem node (or endpoint) from inside and from outside a Session.

**Comment by andre-merzky** [comment]
I am in support of this use case.  Easiest way is to expose [`rp.Session._get_session_sandbox`](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/session.py#L802) and [`rp.Session._get_resource_sandbox`](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/session.py#L677) as part of the public API.  The root filesystem URL is already exposed via [`rp.Session.get_pilot_fs_endpoint](https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/session.py#L853), although that is not helpful as it expects a pilot instance to already exist - we can relax that requirement and use `resource_label` and `access_schema` as parameters.  Potential access schemas are available via `rp.Session.get_resource_config(resource_label)['schemas']`.

In summary, I would propose to expose methods along the following lines:

```py
session = rp.Session()

rcfg = session.get_resource_config('ornl.summit')
schemas = rcfg['schemas'].keys()
root_fs = session.get_fs_url('ornl.summit', 'ssh')
resource_sbox = session.get_resource_sandbox('ornl.summit', 'ssh')
session_sbox = session.get_session_sandbox('ornl.summit', 'ssh')
```

Note that we cannot obtain the pilot sandbox before the pilot exists - the pilot can have a user assigned sandbox and/or a user defined `uid`, and both can change the sandbox path.

**Comment by eirrgang** [comment]
It would be sufficient, but not ideal, if a Session were required, as long as a Pilot is not.

However, the session and pilot sandboxes are runtime details that are actually beyond the scope of the feature request.

I'm hoping simply for a programmatic way to get the value of, say, `rcfg["ssh"]["filesystem_endpoint"] if "ssh" in rcfg["schemas"]`, using the same composition logic that will be employed later, internally (such as for the proposed `endpoint:///` staging scheme).

**Comment by andre-merzky** [comment]
> Could Session.get_resource_config() be a @classmethod? 

In terms of API design: yes.  In terms of implementation: maybe.  The session *instance* is at the moment responsible for loading and parsing the resource config files and we would need to refactor that code quite a bit.  That's possible but quite a bit of effort so I am hesitant to commit.

> Is there any refinement possible in the strength of the specification of the returned data structure?

Absolutely - that's overdue anyway.


> If not, an offline version of `get_fs_endpoint` in the stable API would be very helpful. So would a `job_manager_endpoint` version.

Got your point I think - let me see how best to provide this...

**Comment by andre-merzky** [comment]
After rereading the ticket i realized I may be overthinking this.  The following is readily available and should address some of the use cases.  @eirrgang: would that work at a starting point, and what would be missing from here?

```py
import radical.utils as ru

rcfgs  = ru.Config('radical.pilot.resource', name='*')
js_url = rcfgs.local.localhost.local.job_manager_endpoint

assert js_url == 'fork://localhost/'
```

**Comment by eirrgang** [comment]
> After rereading the ticket i realized I may be overthinking this. The following is readily available and should address some of the use cases. @eirrgang: would that work at a starting point, and what would be missing from here?
> 
> ```python
> import radical.utils as ru
> 
> rcfgs  = ru.Config('radical.pilot.resource', name='*')
> js_url = rcfgs.local.localhost.ssh.job_manager_endpoint 
> 
> assert js_url == 'ssh://localhost/'
> ```

Yes, that is an excellent starting point! The docstring for `ru.Config` is very helpful, too. (Can we get that linked/linkable online somewhere?)

The main thing missing is then the reassurance of what part of the schema is stable or not (or to what degree), if any.

**Comment by andre-merzky** [comment]
> The main thing missing is then the reassurance of what part of the schema is stable or not (or to what degree), if any.

I am not exactly sure what you mean with that.  Do you mean: 'is `ssh` available for all resources`?  Well, maybe not - but the schemas supported by any specific resource can be inspected also:

```py
>>> ...
>>> rcfgs  = ru.Config('radical.pilot.resource', name='*')
>>> print(rcfgs.local.localhost.schemas)
['local', 'ssh']
```

**Comment by andre-merzky** [comment]
I would suggest to close this issue in favor of #2588 which will address the lack of documentation for the resource config.

**Comment by eirrgang** [comment]
> I would suggest to close this issue in favor of #2588 which will address the lack of documentation for the resource config.

#2699, as well, I think. But, yes, the remaining questions in this issue are within scope of other existing issues, I think.

**Comment by eirrgang** [comment]
Ooh! I didn't know I could close issues here. Is that intentional?

**Comment by andre-merzky** [comment]
Yes - you can open then, so you are best qualified to close them :-)  It's our (very loosely followed) policy that PR comments, issues etc. are best resolved / closed by those who open it...
